### PR TITLE
Use JSON for title sequences

### PR DESCRIPTION
**Pull request by ThomasdenH**
Convert the title sequence format to JSON. This will make it easier to expand in the future.

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
While I'm generally all for moving towards standardised formats, I'm not convinced JSON will "make it easier to expand" in and of itself — especially given the existing format is already in plain text.

I'm aware this might make it easier to manage title sequences through JavaScript and other scripting languages, but I don't see how the current format prevents that, either.

What about custom title sequences? It's nice that you've converted the title sequences we ship with, but user-built title sequences will no longer be compatible.

In short, I'm not sure the added overhead of the JSON syntax is worth it. I'd be delighted if you could expand on your reasoning behind these changes, though.

**Comment by IntelOrca** [comment]
@AaronVanGeffen I requested it was changed to JSON, I did it for themes and its consistent with our other files, e.g. groups and users. Its standard, easier to extend and easier to parse.

**Comment by AaronVanGeffen** [comment]
Thanks. That would have been useful background info to have included in the PR description.

What about custom title sequences, though? I'm not too bothered, personally, but it would be nice to retain some kind of compatibility with existing sequences, perhaps.

**Comment by IntelOrca** [comment]
@AaronVanGeffen I don't think there are enough around to warrant it, it shouldn't be too hard or take too long to manually convert.

**Comment by ThomasdenH** [comment]
The current format would mean more and more time for expanding, for example for #4500.

**Comment by janisozaur** [comment]
@ThomasdenH your PR does not compile, can you please fix that?

@spacek531 how does that affect you?

File: data/title/openrct2/script.json [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  { "command": "rotate", "rotations": 3 },
+  { "command": "wait", "seconds": 15},
+  
+  { "command": "location", "x": 100, "y": 63 },
+  { "command": "rotate", "rotations": 2 },
+  { "command": "wait", "seconds": 15 },
+  
+  { "command": "restart" },
+  { "command": "end" }
+]
```
Comment (janisozaur) [inline data/title/openrct2/script.json]: missing newline

File: data/title/rct1/script.json [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  { "command": "location", "x": 60, "y": 62},
+  { "command": "wait", "seconds": 15},
+
+  { "command": "loadrct1", "saveIndex": 18 },
+  { "command": "rotate", "rotations": 3 },
+  { "command": "location", "x": 50, "y": 77},
+  { "command": "wait", "seconds": 15},
+
+  {"command": "restart"}
+]
```
Comment (janisozaur) [inline data/title/rct1/script.json]: missing newline

**Comment by ThomasdenH** [comment]
I checked for warnings, but somehow I didn't see these before. I will fix them.

**Comment by IntelOrca** [comment]
For C++ code you should follow the new C++ conventions we have. Use functions from core/... instead of C functions like malloc and strcpy. Braces on their own lines, spaces instead of tabs, etc. -- compare with other C++ files (apart from mixer and network).

**Comment by spacek531** [comment]
@janisozaur the change seems a little... verbose for what I have seen in the past (where whitespace has been stripped out of the script).

**Comment by ThomasdenH** [comment]
Comments are not possible in JSON unfortunately. I was indeed thinking about a small tool to covert them.

File: src/interface/title_sequences.cpp [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+        try
+        {
+            Json::WriteToFile(path, jsonCommands, JSON_INDENT(4) | JSON_PRESERVE_ORDER);
+        }
+        catch (Exception ex)
+        {
+            log_error("Unable to save title sequence %s: %s", path, ex.GetMessage());
+        }
+    }
+}
```
Comment (janisozaur) [inline src/interface/title_sequences.cpp]: missing newline

File: src/interface/title_sequences.h [hunk 4]
```diff
@@ -35,13 +36,17 @@ title_command TitleScriptMakeCommand(int command, int parameter1, int parameter2
 #define TITLE_LOAD(i)			TitleScriptMakeCommand(TITLE_SCRIPT_LOAD,		i, 0)
 
 // The index of the current title sequence being edited
-extern uint16 gCurrentTitleSequence;
+uint16 gCurrentTitleSequence;
```
Comment (janisozaur) [inline src/interface/title_sequences.h]: that change looks wrong, you're declaring a variable in a header file

File: src/interface/title_sequences.cpp [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+                json_object_set_new(jsonCommand, "command", json_string("end"));
+                break;
+            }
+            json_array_append_new(jsonCommands, jsonCommand);
+        }
+
+        // Write to file
+        try
+        {
+            Json::WriteToFile(path, jsonCommands, JSON_INDENT(4) | JSON_PRESERVE_ORDER);
```
Comment (janisozaur) [inline src/interface/title_sequences.cpp]: you never release `jsonCommands`, I think you should.

File: src/interface/title_sequences.cpp [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+            Memory::CopyArray(gConfigTitleSequences.presets[preset].saves[gConfigTitleSequences.presets[preset].num_saves - 1], fileInfo.path, TITLE_SEQUENCE_MAX_SAVE_LENGTH);
+            gConfigTitleSequences.presets[preset].saves[gConfigTitleSequences.presets[preset].num_saves - 1][TITLE_SEQUENCE_MAX_SAVE_LENGTH - 1] = '\0';
+        }
+        platform_enumerate_files_end(fileEnumHandle);
+
+        // Open the json and parse the sequence
+        try
+        {
+            json_t* json = Json::ReadFromFile(scriptPath);
+            title_sequence_from_json(json, &gConfigTitleSequences.presets[preset]);
```
Comment (janisozaur) [inline src/interface/title_sequences.cpp]: shouldn't you release this json object?

File: src/interface/title_sequences.cpp [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+void title_sequence_from_json(json_t* json, title_sequence* sequence)
+{
+    sequence->num_commands = (uint16) json_array_size(json);
+    sequence->commands = Memory::AllocateArray<title_command>(sequence->num_commands);
+
+    for (int i = 0; i < sequence->num_commands; i++)
+        sequence->commands[i] = title_sequence_command_from_json(json_array_get(json, i), sequence);
+}
+
+extern "C" {
```
Comment (janisozaur) [inline src/interface/title_sequences.cpp]: this is only required at symbol declaration

**Comment by ThomasdenH** [comment]
How do I fix the Xcode project?

**Comment by duncanspumpkin** [comment]
@marijnvdwerf or @LRFLEW can fix the xcode project files

**Comment by LRFLEW** [comment]
@ThomasdenH If you merge ThomasdenH/OpenRCT2#1, it will fix the macOS compilation in the PR.

**Comment by ThomasdenH** [comment]
Thank you!

**Comment by ThomasdenH** [comment]
Can this be merged?

File: src/core/StringBuilder.hpp [hunk 8]
```diff
@@ -96,6 +96,17 @@ class StringBuilder final
     }
 
     /**
+     * Modifies the buffer to be a substring of the current string, beginning at start and ending at end
+     */
+    void SubString(size_t start, size_t end)
```
Comment (janisozaur) [inline src/core/StringBuilder.hpp]: all `substr` implementations return `substr` and are `const`, I'd rather see it done this way here as well.

File: src/interface/title_sequences.cpp [hunk 9]
```diff
@@ -0,0 +1,609 @@
...
+    if (preset >= TITLE_SEQUENCE_DEFAULT_PRESETS && preset < gConfigTitleSequences.num_presets && index > 0 && index < gConfigTitleSequences.presets[preset].num_commands) {
+        title_command command = gConfigTitleSequences.presets[preset].commands[index];
+        gConfigTitleSequences.presets[preset].commands[index] = gConfigTitleSequences.presets[preset].commands[index - 1];
+        gConfigTitleSequences.presets[preset].commands[index - 1] = command;
+        title_sequence_save_preset_script(preset);
+    }
+}
+
+void title_sequence_from_json(json_t* json, title_sequence* sequence);
+title_command title_sequence_command_from_json(json_t* json, title_sequence* sequence);
```
Comment (janisozaur) [inline src/interface/title_sequences.cpp]: this should be removed too.

File: src/interface/title_sequences.cpp [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+            // Remove everything after that
+            StringBuilder* nameBuffer2 = nameBuffer->SubString(0, folderNameEnd);
+            // Find the separator before it
+            size_t folderNameStart = String::LastIndexOf(nameBuffer2->GetBuffer(), *PATH_SEPARATOR);
+            // Remove everything befoe that
+            StringBuilder* nameBuffer3 = nameBuffer2->SubString(folderNameStart + 1, nameBuffer2->GetLength());
+            
+            Memory::CopyArray<char>(gConfigTitleSequences.presets[preset].name, nameBuffer3->GetBuffer(), TITLE_SEQUENCE_NAME_SIZE);
+            gConfigTitleSequences.presets[preset].path[0] = '\0';
+            delete nameBuffer, nameBuffer2, nameBuffer3;
```
Comment (janisozaur) [inline src/interface/title_sequences.cpp]: That's not intended use of comma operator. It does not do what you expect it to.

**Comment by janisozaur** [comment]
@ThomasdenH this still does not compile, are you going to fix it?

**Comment by YJSoft** [comment]
There's no function declation for `title_sequence_from_json`.

**Comment by ThomasdenH** [comment]
Yeah, you asked me to remove them from the .cpp file. I guess I should add them to the header file?

File: data/title/rct1aa/script.txt [hunk 11]
```diff
@@ -1,187 +0,0 @@
...
-LOCATION 97, 72
-WAIT 12
-ROTATE 2
-LOCATION 74, 71
-WAIT 12
-ROTATE 2
-LOCATION 67, 90
-WAIT 12
-
-# SC_GIGGLE_DOWNS
```
Comment (ThomasdenH) [inline data/title/rct1aa/script.txt]: That is true. The main reason to switch is because the parser would get much more complicated when doing issues like #4500 for the current format. I guess one (not ideal) solution would be to create special elements that act like comments and are ignored in code?

**Comment by janisozaur** [comment]
@ThomasdenH is this function meant to be used outside this translation unit?

File: data/title/rct1aa/script.txt (hunk 11 above)
Comment (Gymnasiast) [inline data/title/rct1aa/script.txt]: Make it ignore every parameter it doesn't recognise, and putting comments into a property called "comment" or "//comment" or something like that.

Comment (marijnvdwerf) [inline data/title/rct1aa/script.txt]: Or just pust strings in the array and only attempt to parse objects. That way you don't need a special property.

Comment (IntelOrca) [inline data/title/rct1aa/script.txt]: We can certainly strip comments with a preprocessor, but it makes them unreadable to other applications.

Comment (marijnvdwerf) [inline data/title/rct1aa/script.txt]: Or you could've gone for YAML, which does allow comments.
Another way to fix this would be using strings for the `loadrct1` `saveIndex` property.

Comment (Gymnasiast) [inline data/title/rct1aa/script.txt]: We could still go for YAML, AFAIC.

**Comment by ThomasdenH** [comment]
No, it is not. Should it be static?

**Comment by janisozaur** [comment]
Yes

**Comment by IntelOrca** [comment]
I don't think we need to keep this PR open, the codebase has changed too much. If I ever get round to refactoring all the title sequence code I might think about going to JSON or YAML.

**Comment by ThomasdenH** [comment]
Yes, I might come back to this later this year, but starting over might be quicker at this point.
